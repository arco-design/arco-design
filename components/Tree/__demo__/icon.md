---
order: 11
title: 
  zh-CN: 定制节点图标
  en-US: Node Icon
---

## zh-CN

只需为 `TreeNode` 指定 `icon` 属性的值即可为任意节点指定任意图标。

## en-US

The property `icon` of `TreeNode` can specify an icon for the node.

```js
import { Tree } from '@arco-design/web-react';
import { IconStar } from '@arco-design/web-react/icon';
const TreeNode = Tree.Node;

function App() {
  return (
    <Tree>
      <TreeNode icon={<IconStar />} key="node1" title="Trunk">
        <TreeNode icon={<IconStar />} key="node2" title="Leaf" />
      </TreeNode>
      <TreeNode icon={<IconStar />} key="node3" title="Trunk">
        <TreeNode icon={<IconStar />} key="node4" title="Leaf" />
        <TreeNode icon={<IconStar />} key="node5" title="Leaf" />
      </TreeNode>
    </Tree>
  );
}

export default App;
```
