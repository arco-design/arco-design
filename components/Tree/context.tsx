import { createContext, DragEvent } from 'react';

import { AvailableVirtualListProps } from '../_class/VirtualList';
import { NodeProps, TreeDataType, TreeProps, TreeState } from './interface';

export const TreeContext = createContext<{
  icons?: NodeProps['icons'];
  loadMore?: (node: NodeProps) => void;
  renderExtra?: (node: NodeProps) => void;
  renderTitle?: TreeProps['renderTitle'];
  virtualListProps?: AvailableVirtualListProps;
  onSelect?: (_key: string, e) => void;
  onCheck?: (checked: boolean, _key: string, e) => void;
  onExpand?: (expanded: boolean, _key: string) => void;
  onNodeDrop?: (
    e: DragEvent<HTMLSpanElement>,
    nodeProps: NodeProps,
    dragPosition: 0 | 1 | -1
  ) => void;
  key2nodeProps?: { [key: string]: NodeProps };
  actionOnClick?: TreeProps['actionOnClick'];
  getNodeProps?: <T extends NodeProps | NodeProps[]>(nodes: T) => T;
  getTreeState?: () => TreeState;
  onExpandEnd?: (key: string) => void;
  onNodeDragStart?: (e: DragEvent<HTMLSpanElement>, nodeProps: NodeProps) => void;
  onNodeDragEnd?: (e: DragEvent<HTMLSpanElement>, nodeProps: NodeProps) => void;
  onNodeDragOver?: (e: DragEvent<HTMLSpanElement>, nodeProps: NodeProps, position) => void;
  onNodeDragLeave?: (e: DragEvent<HTMLSpanElement>, nodeProps: NodeProps) => void;
  allowDrop?: (nodeProps: NodeProps, dragPosition: 0 | 1 | -1) => boolean;
  getFieldInfo?: (nodeProps: NodeProps) => TreeDataType;
}>({});
