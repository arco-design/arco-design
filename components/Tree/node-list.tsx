import React, {
  useState,
  useMemo,
  useCallback,
  forwardRef,
  useImperativeHandle,
  useRef,
} from 'react';
import VirtualList, { VirtualListHandle } from '../_class/VirtualList';
import useUpdate from '../_util/hooks/useUpdate';
import Node from './node';
import scrollIntoViewIfNeed from '../_util/scrollIntoView';

function getKey(option) {
  return option.key || option._key;
}

function NodeList(props, ref) {
  const {
    className,
    style,
    filterNode,
    virtualListProps,
    expandedKeys,
    currentExpandKeys,
    saveCacheNode,
    nodeList,
    getNodeProps,
  } = props;
  const isVirtual = virtualListProps?.threshold !== null;
  const virtualListRef = useRef<VirtualListHandle>();
  const treeWrapperRef = useRef<HTMLDivElement>();
  const expandedKeysSet = useMemo(() => new Set(expandedKeys), [expandedKeys]);
  const visibleKeys: Set<string> = useMemo(() => {
    const newKeys = new Set<string>();
    const currentExpandKeysSet = new Set(currentExpandKeys);
    nodeList.forEach((nodeProps) => {
      const pathParentKeys = nodeProps.pathParentKeys || [];
      // 如果父节点处于正在展开状态，子节点暂时不可见，因为父节点的children会在animation中渲染出来。
      // 当动画完成时，父节点children隐藏，此时在这里渲染子节点。 anyway，一切为了动画！！！
      if (
        pathParentKeys.every((key) => !currentExpandKeysSet.has(key) && expandedKeysSet.has(key))
      ) {
        newKeys.add(nodeProps._key);
      }
    });
    return newKeys;
  }, [expandedKeysSet, currentExpandKeys, nodeList]);

  const calcChildrenList = useCallback(() => {
    return nodeList.filter((item) => {
      const pass = !filterNode || (filterNode && filterNode(item));

      if (pass && visibleKeys.has(item.key)) {
        return true;
      }
      // 过滤掉的也缓存一下，避免被收起的节点在onSelect回调中，selectedNodes出现undefined
      saveCacheNode(item);
      return false;
    });
  }, [nodeList, filterNode, visibleKeys]);

  // 默认值不能为nodeList，防止在设置defaultExpandedKeys时，应该被隐藏的节点初始化的时候展示了。
  const [childrenList, setChildrenList] = useState(() => {
    return calcChildrenList();
  });

  useUpdate(() => {
    setChildrenList(calcChildrenList());
  }, [calcChildrenList]);

  useImperativeHandle(ref, () => {
    return {
      // index: 第几个dom元素, 如果传入的是字符串，会作为 node 的 key去查找。
      // nodeProps: _index 是 key 时，对应的node
      scrollIntoView: (_index, nodeProps) => {
        let index = _index;
        const isKey = typeof _index === 'string';

        if (isKey) {
          let key = _index;
          // 查找离得最近的可见的父节点，进行滚动。
          if (!visibleKeys.has(_index) && nodeProps && nodeProps.pathParentKeys) {
            key =
              [...nodeProps.pathParentKeys].reverse().find((key) => visibleKeys.has(key)) || index;
          }
          // _index attributes and index are not the same due to some hidden items
          index = childrenList.findIndex(({ _key }) => _key === key);
        }

        if (!isVirtual && treeWrapperRef.current) {
          const wrapperDom = treeWrapperRef.current;
          const node = wrapperDom ? wrapperDom.children[index] : null;

          node &&
            scrollIntoViewIfNeed(node as HTMLElement, {
              boundary: wrapperDom.parentElement,
            });
        } else if (virtualListRef.current) {
          virtualListRef.current.scrollTo({ index });
        }
      },
    };
  });

  return isVirtual ? (
    <VirtualList
      className={className}
      style={style}
      ref={virtualListRef}
      data={getNodeProps(childrenList)}
      isStaticItemHeight={false}
      itemKey={getKey}
      onMouseDown={props.onMouseDown}
      {...props.ariaProps}
      {...virtualListProps}
    >
      {(item) => {
        const node = <Node {...item} key={item.key} />;
        saveCacheNode(item);
        return node;
      }}
    </VirtualList>
  ) : (
    <div
      role="tree"
      tabIndex={0}
      className={className}
      style={style}
      ref={treeWrapperRef}
      {...props.ariaProps}
      onMouseDown={props.onMouseDown}
    >
      {childrenList.map((item) => {
        const nodeProps = getNodeProps(item);
        const node = <Node {...nodeProps} key={item.key} />;
        saveCacheNode(item);
        return node;
      })}
    </div>
  );
}

export default forwardRef(NodeList);
