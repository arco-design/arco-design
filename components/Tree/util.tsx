import React, { ReactNode, ReactChild } from 'react';
import { TreeDataType } from './interface';
import { key2nodePropsType } from '.';

export const getTreeDataFromTreeChildren = (treeChildren: ReactNode) => {
  const loop = (children) => {
    return React.Children.map(children, (child: ReactChild) => {
      if (!React.isValidElement(child)) {
        return;
      }
      const key = child.key;

      return {
        ...child.props,
        key,
        children: loop(child.props.children),
      };
    });
  };
  return loop(treeChildren);
};

export function getChildNodeKeys(
  node: TreeDataType,
  key2nodeProps: key2nodePropsType
): Set<string> {
  const nodes: Set<string> = new Set();
  const loop = (children: TreeDataType[]) => {
    children.map((child) => {
      const key = child.key;
      const item = key2nodeProps[key];
      if (!item || item.disabled || item.disableCheckbox || item.checkable === false) {
        return;
      }
      nodes.add(key);
      loop(item.children || []);
    });
  };
  if (node) {
    loop(node.children || []);
  }
  return nodes;
}

const updateParent = (key, key2nodeProps, allKeys, indeterminateKeysSet) => {
  const pathParentKeys = [...key2nodeProps[key].pathParentKeys];

  // 逐级更新父节点的状态
  pathParentKeys.reverse().some((itemKey) => {
    const parent = key2nodeProps[itemKey];
    if (parent && !parent.disabled && !parent.disableCheckbox && parent.checkable !== false) {
      let total = 0;
      let number = 0;
      parent.children.some(({ key }) => {
        const item = key2nodeProps[key];
        // 不符合可选条件
        if (!item || item.disabled || item.disableCheckbox || item.checkable === false) {
          return false;
        }
        total++;
        if (allKeys.has(key)) {
          number++;
        } else if (indeterminateKeysSet.has(key)) {
          // 只要有一个半选，就不用再算了 ，父节点是半选
          number += 0.5;
          return true;
        }
      });

      if (!number || number === total) {
        indeterminateKeysSet.delete(itemKey);
      } else {
        indeterminateKeysSet.add(itemKey);
      }

      if (number && number === total) {
        allKeys.add(itemKey);
      } else {
        allKeys.delete(itemKey);
      }
    } else {
      // 断开链接
      return true;
    }
  });
};

// also used by tree-select
export function getCheckedKeysByInitKeys(
  checkedKeys: string[],
  key2nodeProps
): {
  checkedKeys: string[];
  indeterminateKeys: string[];
} {
  const checkedKeysSet = new Set<string>(checkedKeys || []);
  const indeterminateKeysSet = new Set<string>();
  const childCheckedKeysSet = new Set<string>();

  checkedKeys.forEach((key) => {
    if (!childCheckedKeysSet.has(key)) {
      const childKeys = getChildNodeKeys(key2nodeProps[key], key2nodeProps);
      // 选中了节点，就找到所有符合条件的子节点的key.自身的选中状态需要根据children判断。
      childKeys.forEach((v) => {
        childCheckedKeysSet.add(v);
      });
    }

    if (
      key2nodeProps[key] &&
      !key2nodeProps[key].pathParentKeys.some((_key) => checkedKeysSet.has(_key))
    ) {
      updateParent(key, key2nodeProps, checkedKeysSet, indeterminateKeysSet);
    }
  });

  return {
    checkedKeys: [...new Set([...checkedKeysSet, ...childCheckedKeysSet])],
    indeterminateKeys: [...indeterminateKeysSet],
  };
}

// also used by tree-select
export function getAllCheckedKeysByCheck(
  key: string,
  checked: boolean,
  checkedKeys: string[],
  key2nodeProps: key2nodePropsType,
  indeterminateKeys: string[]
): {
  checkedKeys: string[];
  indeterminateKeys: string[];
} {
  if (!key2nodeProps[key]) {
    return {
      checkedKeys,
      indeterminateKeys,
    };
  }
  const checkedKeysSet = new Set(checkedKeys);
  const indeterminateKeysSet = new Set(indeterminateKeys);
  const childKeys = getChildNodeKeys(key2nodeProps[key], key2nodeProps);
  const allKeys = checkedKeysSet;

  if (checked) {
    // 选中了节点，就找到所有符合条件的子节点的key.自身的选中状态需要根据children判断。
    allKeys.add(key);
    indeterminateKeysSet.delete(key);
    childKeys.forEach((v) => {
      allKeys.add(v);
    });
  } else {
    indeterminateKeysSet.delete(key);
    // 移除所有符合条件的子节点的key
    allKeys.delete(key);
    childKeys.forEach((v) => {
      allKeys.delete(v);
    });
  }

  // 之后逐级更新父节点的选中状态。
  updateParent(key, key2nodeProps, checkedKeysSet, indeterminateKeysSet);

  return {
    checkedKeys: [...allKeys],
    indeterminateKeys: [...indeterminateKeysSet],
  };
}
