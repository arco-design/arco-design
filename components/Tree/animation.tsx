/**
 * 该组件用来切换tree 展开收起时的动画
 */

import React, { PropsWithChildren, useMemo, useContext, useEffect } from 'react';

import { CSSTransition } from 'react-transition-group';
import { TreeContext } from './context';
import { NodeProps } from './interface';
import VirtualList from '../_class/VirtualList';
import { ConfigContext } from '../ConfigProvider';
import Node from './node';
import { isNumber } from '../_util/is';

function getKey(option) {
  return option.key || option._key;
}

const TreeAnimation = (props: PropsWithChildren<NodeProps>) => {
  const treeContext = useContext(TreeContext);
  const { getPrefixCls } = useContext(ConfigContext);
  const prefixCls = getPrefixCls('tree-node');
  const { expandedKeys, currentExpandKeys } = treeContext.getTreeState();
  const expanded = props.expanded;

  useEffect(() => {
    return () => {
      treeContext.onExpandEnd && treeContext.onExpandEnd(props._key);
    };
  }, []);

  const childrenPropsList = useMemo(() => {
    const result = [];
    const loop = (list) => {
      list.forEach((item) => {
        const data = treeContext.getFieldInfo(item);
        result.push(data);
        if (data.children && data.children.length) {
          loop(data.children);
        }
      });
    };

    loop(props.childrenData || []);
    return result;
  }, [props.childrenData]);

  const filtedData = useMemo(() => {
    const result = [];
    if (childrenPropsList.length) {
      const expandedKeysSet = new Set(expandedKeys || []);
      childrenPropsList.forEach((data) => {
        let isShow;
        const itemProps = {
          ...treeContext.key2nodeProps[data.key],
        };
        if (expanded) {
          // 只有在每一个父节点都是展开状态时，自己才会展示出来
          isShow =
            itemProps.parentKey === props._key ||
            itemProps.pathParentKeys?.every((key) => {
              return expandedKeysSet.has(key);
            });
        } else if (itemProps.pathParentKeys) {
          // 收起时，只有在props._key 对应的位置之后的所有的自己的父节点都是展开状态，才会展示自己
          const index = itemProps.pathParentKeys.indexOf(props._key);

          isShow = itemProps.pathParentKeys.slice(index + 1).every((key) => {
            return expandedKeysSet.has(key);
          });
        }
        if (isShow) {
          result.push({
            ...itemProps,
            ...treeContext.getNodeProps(itemProps, expandedKeysSet),
            key: data.key,
          });
        }
      });
    }
    return result;
  }, [childrenPropsList, props._key, expanded]);

  let realHeight = treeContext.virtualListProps?.height;
  realHeight = isNumber(realHeight) ? realHeight : 0;

  useEffect(() => {
    // node set loadingMore but has no child nodes.
    // Animation will not be triggered and needs to be removed manually
    if (currentExpandKeys.indexOf(props._key) > -1 && filtedData.length === 0) {
      treeContext.onExpandEnd(props._key);
    }
  }, [filtedData, currentExpandKeys]);

  return (
    <CSSTransition
      in={currentExpandKeys.indexOf(props._key) > -1 && filtedData.length > 0}
      unmountOnExit
      classNames="tree-slide-expand"
      timeout={{
        enter: 200,
        exit: 0,
      }}
      onEnter={(e) => {
        const scrollHeight = e.scrollHeight;
        e.style.height = expanded ? 0 : `${Math.min(realHeight || scrollHeight, e.scrollHeight)}px`;
      }}
      onEntering={(e) => {
        const scrollHeight = e.scrollHeight;
        e.style.height = expanded ? `${Math.min(realHeight || scrollHeight, scrollHeight)}px` : 0;
      }}
      onEntered={(e) => {
        e.style.height = props.expanded ? '' : 0;
        treeContext.onExpandEnd(props._key);
      }}
      onExit={(e) => {
        e.style.display = 'none';
      }}
    >
      <VirtualList
        itemKey={getKey}
        className={`${prefixCls}-list`}
        isStaticItemHeight={false}
        {...treeContext.virtualListProps}
        data={filtedData}
        style={{ overflow: 'hidden' }}
      >
        {(child) => {
          return <Node {...child} />;
        }}
      </VirtualList>
    </CSSTransition>
  );
};

export default TreeAnimation;
