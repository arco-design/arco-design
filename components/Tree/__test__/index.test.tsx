import React from 'react';
import { render, fireEvent, act } from '../../../tests/util';
import mountTest from '../../../tests/mountTest';
import Tree from '..';
import { IconHeartFill } from '../../../icon';
import componentConfigTest from '../../../tests/componentConfigTest';

componentConfigTest(Tree, 'Tree');

mountTest(Tree);
const TreeNode = Tree.Node;
const prefixCls = '.arco-tree';

const getNodesContent = (nodes) => {
  return [].slice.call(nodes).map((x) => x.querySelector('.arco-tree-node-title-text').textContent);
};

const getCheckedKeys = () => {
  return [].slice
    .call(document.querySelectorAll('.arco-checkbox-checked'))
    .map((x) => x.querySelector('input').getAttribute('value'));
};

const data = [
  {
    key: 'node1',
    title: '拉尼斯特家族',
    children: [
      {
        key: 'node1-1',
        title: '小恶魔',
        children: [
          {
            key: 'node1-1-1',
            title: '小小恶魔',
          },
        ],
      },
    ],
  },
  {
    key: 'node2',
    title: '史塔克家族',
    children: [
      {
        key: 'node2-1',
        title: '二丫',
      },
      {
        key: 'node2-2',
        title: '三傻',
      },
    ],
  },
];

const defaultSelectedKeys = ['node2'];
const defaultExpandedKeys = ['node1'];

// 从treedata 生成 treenode
const generatorTreeNodes = (treeData) => {
  return treeData.map((item) => {
    const { children, key, ...rest } = item;
    return (
      <Tree.Node key={key} {...rest} dataRef={item}>
        {children ? generatorTreeNodes(item.children) : null}
      </Tree.Node>
    );
  });
};

const getTreeNodesLength = (treeData) => {
  let results = treeData.length;
  for (let i = 0; i < treeData.length; i++) {
    const item = treeData[i];
    const { children } = item;
    if (children) {
      results += getTreeNodesLength(children);
    } else {
      continue;
    }
  }
  return results;
};

// const getLeafLength = (treeData) => {
//   const obj = Array.isArray(treeData) ? { children: treeData } : treeData;
//   if (!obj.children || obj.children.length === 0) {
//     return 1;
//   }
//   let leafCount = 0;
//   for (let i = 0; i < obj.children.length; i++) {
//     leafCount += getLeafLength(obj.children[i]);
//   }

//   return leafCount;
// };
describe('Tree', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runAllTimers();
  });

  it('tree render basic', () => {
    const wrapper = render(<Tree>{generatorTreeNodes(data)}</Tree>);
    const summaryNodeLength = getTreeNodesLength(data);
    expect(wrapper.find(`${prefixCls}-node`)).toHaveLength(summaryNodeLength);
  });
  it('tree render treedata', () => {
    const wrapper = render(<Tree treeData={data} />);
    const wrapper2 = render(<Tree>{generatorTreeNodes(data)}</Tree>);
    expect(wrapper.querySelector('.arco-tree')?.innerHTML).toEqual(
      wrapper2.querySelector('.arco-tree')?.innerHTML
    );
  });

  it('should be expanded and collapsed', async () => {
    const mockExpand = jest.fn();
    const wrapper = render(
      <Tree defaultExpandedKeys={defaultExpandedKeys} onExpand={mockExpand}>
        {generatorTreeNodes(data)}
      </Tree>
    );
    expect(wrapper.find('.arco-tree-node').item(0)).toHaveClass('arco-tree-node-expanded');

    // 收起 node1
    fireEvent.click(wrapper.find(`${prefixCls}-node-switcher-icon`).item(0));

    jest.runAllTimers();

    expect(
      wrapper.find('.arco-tree-node').item(0).classList.contains('arco-tree-node-expanded')
    ).toBeFalsy();

    // onExpand无法触发的bug
    expect(mockExpand.mock.calls.length).toBe(1);

    const nodes = wrapper.find(`${prefixCls}-node-switcher-icon`);

    expect(nodes.length).toBe(2);

    // 展开 node2
    fireEvent.click(nodes[1]);

    jest.runAllTimers();

    expect(wrapper.find('.arco-tree-node').item(1)).toHaveClass('arco-tree-node-expanded');
  });

  it('should be expanded and collapsed when expandedKeys in props', async () => {
    let keys: string[] = ['node1-1-1'];
    const wrapper = render(
      <Tree expandedKeys={keys} onExpand={(v) => (keys = v)}>
        {generatorTreeNodes(data)}
      </Tree>
    );

    // 子节点展开，子节点的父节点也需要都展开
    expect(wrapper.querySelectorAll('[aria-expanded=true]')).toHaveLength(3);

    expect(getNodesContent(wrapper.querySelectorAll('[aria-expanded=true]'))).toEqual([
      '拉尼斯特家族',
      '小恶魔',
      '小小恶魔',
    ]);
    fireEvent.click(wrapper.find(`${prefixCls}-node-switcher-icon`).item(1));

    // 父节点关闭，子节点保持展开
    expect(keys).toEqual(['node1-1-1', 'node1']);

    wrapper.rerender(
      <Tree expandedKeys={keys} onExpand={(v) => (keys = v)}>
        {generatorTreeNodes(data)}
      </Tree>
    );

    wrapper.rerender(
      <Tree expandedKeys={['node2-1', 'node2']} onExpand={(v) => (keys = v)}>
        {generatorTreeNodes(data)}
      </Tree>
    );

    expect(wrapper.find('.arco-tree-node-expanded')).toHaveLength(2);
    expect(getNodesContent(wrapper.querySelectorAll('.arco-tree-node-expanded'))).toEqual([
      '史塔克家族',
      '二丫',
    ]);
  });

  it('should be selected correctly', () => {
    const mockSelected = jest.fn();

    const wrapper = render(
      <Tree defaultSelectedKeys={defaultSelectedKeys} onSelect={mockSelected}>
        {generatorTreeNodes(data)}
      </Tree>
    );

    expect(wrapper.find('.arco-tree-node-selected')).toHaveLength(1);

    expect(wrapper.find('.arco-tree-node').item(3)).toHaveClass('arco-tree-node-selected');
    fireEvent.click(wrapper.find(`${prefixCls}-node-title`).item(0));
    expect(mockSelected.mock.calls).toHaveLength(1);

    expect(wrapper.find('.arco-tree-node-selected')).toHaveLength(1);
    expect(wrapper.find('.arco-tree-node').item(0)).toHaveClass('arco-tree-node-selected');
  });
  it('should be selected correctly when multiple', () => {
    const mockSelected = jest.fn();
    const wrapper = render(
      <Tree multiple defaultSelectedKeys={defaultSelectedKeys} onSelect={mockSelected}>
        {generatorTreeNodes(data)}
      </Tree>
    );
    expect(wrapper.find('.arco-tree-node-selected')).toHaveLength(1);

    expect(wrapper.find('.arco-tree-node').item(3)).toHaveClass('arco-tree-node-selected');

    fireEvent.click(wrapper.find(`${prefixCls}-node-title`).item(0));
    expect(mockSelected.mock.calls).toHaveLength(1);

    expect(wrapper.find('.arco-tree-node-selected')).toHaveLength(2);
    expect(wrapper.find('.arco-tree-node').item(0)).toHaveClass('arco-tree-node-selected');
    expect(wrapper.find('.arco-tree-node').item(3)).toHaveClass('arco-tree-node-selected');
  });

  it('should checked correctly', async () => {
    const defaultCheckedKeys = ['node1'];
    let checkedKeys: string[] = [];
    const wrapper = render(
      <Tree
        checkable
        onCheck={(keys) => {
          checkedKeys = keys;
        }}
        defaultCheckedKeys={defaultCheckedKeys}
      >
        {generatorTreeNodes(data)}
      </Tree>
    );

    expect(
      [].slice.call(wrapper.querySelectorAll('input[checked]')).map((x) => x.getAttribute('value'))
    ).toEqual(['node1', 'node1-1', 'node1-1-1']);

    fireEvent.click(
      wrapper.find(`${prefixCls}-node`).item(3).querySelector(`.arco-checkbox`) as any
    );

    expect(checkedKeys).toEqual(['node1', 'node1-1', 'node1-1-1', 'node2', 'node2-1', 'node2-2']);

    expect(getCheckedKeys()).toEqual([
      'node1',
      'node1-1',
      'node1-1-1',
      'node2',
      'node2-1',
      'node2-2',
    ]);
  });

  it('should checkStrictly correctly', () => {
    const mockChecked = jest.fn();
    const defaultCheckedKeys = ['node1'];
    const wrapper = render(
      <Tree checkable checkStrictly onCheck={mockChecked} defaultCheckedKeys={defaultCheckedKeys}>
        {generatorTreeNodes(data)}
      </Tree>
    );

    // 目前通过
    expect(getCheckedKeys()).toEqual(['node1']);

    act(() => {
      fireEvent.click(
        wrapper.find(`${prefixCls}-node`).item(3).querySelector(`.arco-checkbox`) as any
      );
    });

    expect(mockChecked.mock.calls).toHaveLength(1);

    expect(getCheckedKeys()).toEqual(['node1', 'node2']);
  });

  it('should render empty correctly', () => {
    const wrapper = render(<Tree>{generatorTreeNodes([])}</Tree>);
    expect(wrapper.find('.arco-tree')).toHaveLength(1);
    expect(wrapper.find(`${prefixCls}-node`)).toHaveLength(0);
  });

  it('should loadMore correctly', () => {
    const loadMore = jest.fn();
    const wrapper = render(<Tree loadMore={loadMore}>{generatorTreeNodes(data)}</Tree>);
    const iconLength = wrapper.find(`${prefixCls}-node-switcher-icon`).length;
    expect(wrapper.find(`${prefixCls}-node`).length).toEqual(iconLength);
    const nodes = wrapper.find(`${prefixCls}-node-switcher-icon`);
    fireEvent.click(nodes[nodes.length - 1]);
    expect(loadMore.mock.calls).toHaveLength(1);
  });

  it('selected correctly when disabled', () => {
    const mockSelected = jest.fn();
    const wrapper = render(
      <Tree onSelect={mockSelected}>
        <TreeNode key="node1" title="拉尼斯特家族" disabled>
          <TreeNode key="node2" title="小恶魔">
            <TreeNode key="node21" title="小恶魔aaa" />
          </TreeNode>
        </TreeNode>
        <TreeNode key="node3" title="史塔克家族">
          <TreeNode key="node4" title="二丫" />
          <TreeNode key="node5" title="三傻" />
        </TreeNode>
      </Tree>
    );

    const disabledItem = wrapper.find(`${prefixCls}-node-disabled`);
    expect(disabledItem).toHaveLength(1);
    fireEvent.click(disabledItem.item(0).querySelector(`${prefixCls}-node-title`) as any);
    expect(mockSelected.mock.calls).toHaveLength(0);
  });

  it('self icon render correctly', () => {
    const wrapper = render(
      <Tree>
        <TreeNode key="node1" title="拉尼斯特家族" icon={<IconHeartFill />} />
        <TreeNode key="node2" title="史塔克家族" />
      </Tree>
    );

    expect(wrapper.find('.arco-icon-heart-fill')).toHaveLength(1);
    //  expect(wrapper.find(`${prefixCls}-node-switcher-icon`)).toHaveLength(1);
  });

  it('should selectable, disableCheckbox correctly', () => {
    const mockSelected = jest.fn();
    const mockChecked = jest.fn();
    const wrapper = render(
      <Tree
        checkable
        defaultSelectedKeys={defaultSelectedKeys}
        onSelect={mockSelected}
        onCheck={mockChecked}
        treeData={[
          ...data,
          {
            key: 'disbaled',
            label: 'disablde',
            selectable: false,
            disableCheckbox: true,
          },
        ]}
      />
    );
    expect(wrapper.find('.arco-tree-node-selected')).toHaveLength(1);

    expect(wrapper.find('.arco-tree-node').item(3)).toHaveClass('arco-tree-node-selected');

    const nodes = wrapper.find(`${prefixCls}-node`);

    fireEvent.click(nodes[nodes.length - 1].querySelector(`${prefixCls}-node-title`) as any);
    fireEvent.click(nodes[nodes.length - 1].querySelector(`.arco-checkbox`) as any);

    expect(mockSelected.mock.calls).toHaveLength(0);
    expect(mockChecked.mock.calls).toHaveLength(0);
    expect(wrapper.find('.arco-tree-node-selected')).toHaveLength(1);

    expect(wrapper.find('.arco-tree-node').item(3)).toHaveClass('arco-tree-node-selected');
  });

  it('should checkedStrategy correctly', async () => {
    let checkedKeys = ['node1'];
    const wrapper = render(
      <Tree
        checkable
        checkedStrategy={Tree.SHOW_PARENT}
        onCheck={(value) => {
          checkedKeys = value;
        }}
        checkedKeys={checkedKeys}
      >
        {generatorTreeNodes(data)}
      </Tree>
    );

    const rerender = (props) => {
      wrapper.rerender(
        <Tree
          checkable
          checkedStrategy={Tree.SHOW_PARENT}
          onCheck={(value) => {
            checkedKeys = value;
          }}
          checkedKeys={checkedKeys}
          {...props}
        >
          {generatorTreeNodes(data)}
        </Tree>
      );
    };

    expect(getCheckedKeys()).toEqual(['node1', 'node1-1', 'node1-1-1']);

    fireEvent.click(wrapper.find(`.arco-checkbox`).item(3));
    expect(checkedKeys).toEqual(['node1', 'node2']);

    rerender({ checkedKeys });

    // jest.runAllTimers();

    expect(getCheckedKeys()).toEqual([
      'node1',
      'node1-1',
      'node1-1-1',
      'node2',
      'node2-1',
      'node2-2',
    ]);

    rerender({ checkedStrategy: Tree.SHOW_CHILD });

    // 取消 node2 选中
    fireEvent.click(wrapper.find(`.arco-checkbox`).item(3));

    expect(checkedKeys).toEqual(['node1-1-1']);

    rerender({ checkedKeys });

    expect(getCheckedKeys()).toEqual(['node1', 'node1-1', 'node1-1-1']);

    rerender({ checkStrictly: true });

    // 选中 node2
    fireEvent.click(wrapper.find(`.arco-checkbox`).item(3));
    expect(checkedKeys).toEqual(['node1-1-1', 'node2']);

    rerender({ checkedKeys, checkStrictly: true });

    expect(getCheckedKeys()).toEqual(checkedKeys);

    expect(wrapper.find('.arco-checkbox.arco-checkbox-checked')).toHaveLength(2);
  });

  it('should halfchecked correctly', () => {
    const defaultCheckedKeys = ['node1'];
    const wrapper = render(
      <Tree checkStrictly checkable halfCheckedKeys={defaultCheckedKeys}>
        {generatorTreeNodes(data)}
      </Tree>
    );

    expect(wrapper.find(`${prefixCls}-node`).item(0).querySelector(`.arco-checkbox`)).toHaveClass(
      'arco-checkbox-indeterminate'
    );
  });

  it('should be expanded when click node', () => {
    const mockExpand = jest.fn();
    const mockSelected = jest.fn();
    const wrapper = render(
      <Tree
        defaultExpandedKeys={[]}
        actionOnClick="expand"
        onSelect={mockSelected}
        onExpand={mockExpand}
      >
        {generatorTreeNodes(data)}
      </Tree>
    );

    expect(wrapper.find('.arco-tree-node-title-text').item(0).textContent).toBe('拉尼斯特家族');

    fireEvent.click(wrapper.find(`${prefixCls}-node-title`).item(0));
    expect(mockExpand.mock.calls).toHaveLength(1);
    expect(mockSelected.mock.calls).toHaveLength(0);
    expect(wrapper.find('.arco-tree-node-title-text').item(1).textContent).toBe('小恶魔');
  });

  it('should be checked and expand when click node', () => {
    let checkedKeys;
    const wrapper = render(
      <Tree
        defaultExpandedKeys={[]}
        checkable
        onCheck={(keys) => (checkedKeys = keys)}
        actionOnClick={['check', 'expand']}
      >
        {generatorTreeNodes(data)}
      </Tree>
    );

    expect(wrapper.find('.arco-tree-node-title-text').item(0).textContent).toBe('拉尼斯特家族');
    fireEvent.click(wrapper.find(`${prefixCls}-node-title`).item(0));
    expect(wrapper.find('.arco-tree-node-title-text').item(1).textContent).toBe('小恶魔');
    expect(wrapper.find('.arco-checkbox-checked')).toHaveLength(2);

    expect(checkedKeys).toEqual(['node1', 'node1-1', 'node1-1-1']);
  });
});
