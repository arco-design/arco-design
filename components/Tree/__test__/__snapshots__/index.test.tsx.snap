// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Tree default 1`] = `
<div
  class="arco-tree"
  role="tree"
  tabindex="0"
/>
`;

exports[`ConfigProvider componentConfig.Tree set className globally 1`] = `
<div
  class="arco-tree global-default-class"
  role="tree"
  tabindex="0"
/>
`;

exports[`ConfigProvider componentConfig.Tree set className globally and component set className itself. 1`] = `
<div
  class="arco-tree component-class"
  role="tree"
  tabindex="0"
/>
`;

exports[`ConfigProvider componentConfig.Tree set data-* property 1`] = `
<div
  class="arco-tree"
  data-test-name="Tree"
  role="tree"
  tabindex="0"
/>
`;
